source/develop/release-management/features/storage/multipath-events.html.md

```diff
@@ -0,0 +1,167 @@
+---
+title: Multipath Events
+category: feature
+authors: Fred Rolland
+feature_name: Multipath Events
+feature_modules: engine,vdsm
+feature_status: Design
+---
+
+# Multipath Events
+
+
+## Overview
+
+This feature adds the ability to display events to the user in
+the engine UI when multipath device path changes its status.
+The supported events are when a path moves to down or up state.
+
+
+## Owner
+
+- Name: [Fred Rolland](https://github.com/rollandf)
+- Email: <frolland@redhat.com>
+
+
+## Detailed Description
+
+
+### How it works
+
+When a host is connected to a DC (Storage Pool), Vdsm will register
+to udev multipath events. For each event, a message will be sent to Engine
+with the details of the event. Engine will create an audit log event. 
+Once the host is disconnected from the DC, Vdsm will unregister from
+the 'udev' events.
+
+
+### New vdsm API
+
+New multipath event type
+
+
+#### Multipath Event
+
+An event with information about the multipath device, path and status.
+
+A mapping with these keys:
+- Type: Path Failed/Path Reinstated (DM_ACTION)
+- Device (GUID): unique GUID of this device (DM_NAME)
+- TBD
+
+
+### Vdsm event monitoring
+
+A new framework in Vdsm is introduced to register to events:
+- A listener can register to a topic, providing a callback method/object.
+- A topic monitor will be created with the first listener.
+- Once the monitor receives an event, it will call the listener callback
+(in a different thread than the one the listener registered)
+- A listener can unregister from a topic.
+- The topic monitor will stop monitoring once the last listener unregister.
```

Comment:
I think we can simplify this to:

- Code interested in receiving mutliapth events will subscribe a callback to receive the event
- when an event is received, all subscribers will receive the events (in another thread).
- If nobody is registered for events, the events are dropped
- The multipath event montior will be started when connecting to the pool
- The multipath event monitor will be stopped when disconnecting from the pool

Reply:
Do we listen to events of multipath devices that we have not connected/disconnected? (I think we should)

Reply:
done

Reply:
@mykaul  yes, we will listen to events on all multipath devices